Add Ubuntu 22.04 builds

slarse opened the pull request:
## 🧢 Changes
Replaces the default Ubuntu 24.04 build with a Ubuntu 22.04 build.

The Ubuntu 24.04 build is hidden away under a different S3 object key. Depending on how the access control is configured on the S3 bucket, configuration may be needed to access the Ubuntu 24.04 build.

<details>
<summary>The original reasoning of this post (pretty outdated now)</summary>
This PR does two (highly related) things.

1. Adds Ubuntu 22.04 builds for x86_64 and ARM64 (side note: are the Linux ARM64 builds actually used??)
2. ~Upgrades the core Linux dependencies, most notably libwebkit2gtk-4.1~
    - Due to large forward-compatibility issues, we'll take this separately

I've been fiddling around with this in my fork to try stuff out in the actual runners, and produced [the Ubuntu 22.04 and Ubuntu 24.04 builds found here](https://github.com/slarse/gitbutler/actions/runs/19802219644).

> Note: Don't mind the fact that the artifacts are 500 MB each, I was veeery lazy and just added the entire `target/release/bundle` directory as an artifact. In other words, the artifacts contain way more stuff than they need to.

The Ubuntu 24.04 AppImage appears to be working very well on my Arch Linux system under Wayland.

I was hoping the Ubuntu 22.04 AppImage would also work, but I'm having some very strange linking issues with it when running on my Arch machine (see below). So, I do believe we need to build for both older and newer systems, independently.

<details>
<summary>Possible linker error when running Ubuntu 22.04 AppImage on Arch Linux</summary>

I run into this error when trying to work with remotes.

```bash
command: fetch_from_remotes
params: {"projectId":"ff91a7a9-1b3d-4b71-abee-6254bc1b7731","action":"modal"})

backend error: git command exited with non-zero exit code 128: ["fetch", "--quiet", "--prune", "upstream", "+refs/heads/*:refs/remotes/upstream/*"]

STDOUT:


STDERR:
git: /tmp/.mount_GitButGDbenp/usr/lib/libpcre2-8.so.0: no version information available (required by git)
/usr/lib/git-core/git: /tmp/.mount_GitButGDbenp/usr/lib/libpcre2-8.so.0: no version information available (required by /usr/lib/git-core/git)
/usr/lib/git-core/git-remote-https: symbol lookup error: /usr/lib/libcurl.so.4: undefined symbol: nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation
fatal: remote helper 'https' aborted session
```

The strangeness is that `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation` is absolutely defined in `/usr/lib/libcurl.so.4`, so there has to be some linker issue. This error does not occur when using the Ubuntu 24.04 AppImage.
</details>

## ☕️ Reasoning

[See this comment](https://github.com/gitbutlerapp/gitbutler/issues/7671#issuecomment-3591558807) for a pretty comprehensive justification for the addition of the Ubuntu 22.04 build.

As for the dependency upgrades, they are well overdue. The WebkitGTK build that's currently used is years old and pretty buggy. Upgrading the dependencies fixes the freeze issue described in #9955 on my Arch Linux machine, and probably touches many of the other issues reported on the Linux bundles, the AppImage in particular.

## 🎫 Affected issues

Fixes: #9955 (The dependency upgrades in the Ubuntu 24.04 image seems to eliminate this problem)

Also improves upon #7671 with the Ubuntu 22.04 build, which links to GLIBC 2.35 instead of 2.38 (as is the case in the Ubuntu 24.04 build). This will let the AppImage run on older distros.

Very likely touches upon more of the Linux-specific issues, but I have not had the time to evaluate more of them.

## TODO
On 1/3 systems I've tested on (another Arch Linux+Wayland machine), I'm getting `Could not create default EGL display: EGL_BAD_PARAMETER. Aborting ...`. Which is strange. I want to figure that out before proceeding here.

### Testing matrix (AppImage)

| Machine | AppImage | Status | Workaround |
| ------------- | --------------- | --------- | ------------------ |
| Arch Linux + Wayland (mesa) | Ubuntu 22.04 | :x: EGL_BAD_PARAMETER | https://github.com/gitbutlerapp/gitbutler/issues/5282#issuecomment-3598600651 (poor xwayland performance) |
| Arch Linux + Wayland (mesa) | Ubuntu 24.04 | :x: EGL_BAD_PARAMETER | https://github.com/gitbutlerapp/gitbutler/issues/5282#issuecomment-3598600651 (poor xwayland performance) |
| Arch Linux + Wayland (Nvidia) | Ubuntu 22.04 | :warning: Cannot work with remotes ||
| Arch Linux + Wayland (Nvidia) | Ubuntu 24.04 | :heavy_check_mark: No issues detected| |
| Ubuntu 22.04 + X11 (mesa) | Ubuntu 22.04 | :heavy_check_mark: No issues detected | |
| Ubuntu 22.04 + X11 (mesa) | Ubuntu 24.04 | :x: Cannot launch due to GLIBC mismatch | |

</detalis>

## Important caveat: The Ubuntu 22.04 build may not last
I don't know for how long the Ubuntu 22.04 build will keep on building, as the WebKit project dropped official support for it [on 2025-04-25](https://docs.webkit.org/Ports/WebKitGTK%20and%20WPE%20WebKit/DependenciesPolicy.html). Which is odd as Debian Bookworm (the Ubuntu 22.04 base distro) has support until June 2026. `libwebkit2gtk-4.1` is however [still an officially Ubuntu-maintained package](https://packages.ubuntu.com/jammy/libwebkit2gtk-4.1-dev), and the current upstream version 2.50.1 [was released in October](https://webkitgtk.org/2025/10/10/webkitgtk2.50.1-released.html). I think that's sufficient recent activity to make it worthwhile to start running the build.

Byron commented:
Thanks so much, this PR could be very impactful to the linux users out there!

Without much ado, I am creating a new [nightly](https://github.com/gitbutlerapp/gitbutler/actions/runs/19814434277) from it and invite you to test out the official builds, as downloaded from the main site.

Something I am wondering about is if the additional ubuntu 22 images will confuse the website. And even if not, I think we will need an upgrade so these builds can be differentiated.

> (side note: are the Linux ARM64 builds actually used??)

They are downloadable, but… I am sure I am missing something.

---- 

And it turns out that the build is failing: https://github.com/gitbutlerapp/gitbutler/actions/runs/19814434277/job/56762812224 .

My feeling is that it *might* re-use caches from other ubuntu versions as the shared key only contains the target, and maybe that causes trouble. The failed ubuntu 22 job used `v0-rust-app-release-build-nightly-Linux-x64-341760e0-379a590c`, and the working ubuntu 24 job used `v0-rust-app-release-build-nightly-Linux-x64-341760e0-379a590c` as well. If nothing else, it's probably racy when writing back. But is this really the cause of the failure? I don't know. But I do know that the weirdest errors happen if the Cargo cache is borked, something that regularly happens even when switching compiler versions. And maybe that's the real issue here.

```
 --> crates/but-claude/src/notifications.rs:3:5
  |
3 | use notify_rust::Notification;
  |     ^^^^^^^^^^^ can't find crate

   Compiling gitbutler-edit-mode v0.0.0 (/home/runner/work/gitbutler/gitbutler/crates/gitbutler-edit-mode)
For more information about this error, try `rustc --explain E0463`.
error: could not compile `but-claude` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
failed to build app: failed to build app
       Error [tauri_cli_node] failed to build app: failed to build app
 ELIFECYCLE  Command failed with exit code 1.
```

Let me try that.

Byron commented:
> Darn, I think I win this year's miscommunication award. I'm working full time right now (not on GitButler :( ), and then I have family responsibilities, so work on GB is generally confined to late evenings and weekends. So I didn't have time to pull down those build artifacts :(. Just quickly popped my head in when my work computer had the EGL_BAD_PARAMETER issue.

No worries, at all! It's good that the binaries can still be downloaded, they have just be superseded by a more recent version.
So I don't think any time was wasted (and even if some was, that's life :D).

> Not entirely sure what we do in the mean time. I think adding the Ubuntu 22.04 build would be worthwhile regardless, but I'm a bit unsure about bumping the core dependencies given the amount of trouble it causes. Then again, _not_ bumping the deps also causes a whole bunch of random issues. Perhaps the preload workaround is fine for now, and we could always just roll back if _that_ causes too many issues for people.
> 
> What do you think?

Actually I can't really judge if the trade-off is worth it. To me right now it *looks* like the old dependencies allow the app to launch 'more often', but it may freeze later, and the new dependency version prevent a launch more often unless the `LD_PRELOAD` workaround is used, but it fixes a bunch of random issues as well.

To me neither seems great and I'd love to downgrade linux support to 'experimental' or something. When that's done, we could move to the latest greatest, have a documentation page with workarounds and make sure people know about it, while waiting for https://github.com/tauri-apps/tauri/pull/12491 to land in the hopes for the one good bundle we can provide.

slarse commented:
> Hey!
> 
> I just wanted to hop in here to go over what went awry with the downloads APIs.
> 
> Our API provides a `platform` property (known as `platform_human` internally). When generating this, it first tried to get a "os_human" which it concatenated with another string. This `os_human` expected the `os` to be either "windows", "darwin", or "linux", so the "linux-..." values caused `os_human` to return `nil` and subsequently fail when we tried to concatenate that into a string.

Thanks for the lore! My third PR causing some kind of major problem is a new personal record :)

> It's our fault for not having any validation on what we expected the `os` value to be. I've made a PR that adds in appropriate validations & gracefully handles the cases when unexpected data is present. I'll be deploying this on Monday.

I prefer not to talk about fault, it's a precarious thing, even when self-assigned. The context of the "faulty" decision is almost always missing from the equation. We break things, we fix things, we move on. Sorry about the sermon :)

But good on the fix for Monday!

> It does however make me wonder how we should go about labelling these different builds. For our different linux builds in the past they've been part of the file name.

I don't think it's a particularly significant decision as the amount of data is so tiny and can easily be reordered later with a small data migration. But I would personally categorize different Linux distributions as different operating systems. Ubuntu 22.04 and Ubuntu 24.04 are different OSes in much the same sense that Windows 10 and Windows 11 are. Typically that which is built for the predecessor works on the successor (but not always), and what's built on the successor does not work on the predecessor (but sometimes it does).

> We could make the required changes to the backend & downloads page to support new OS variants.

Let's not put any effort into this right now, as the Linux situation is a bit fluid. We're charting out a path that entails solid support for Ubuntu 22.04 and 24.04, but it hasn't entirely materialized yet, so effort should be kept to a minimum. This PR has well and truly spiraled beyond all intentions.

If it's possible to keep pushing both Ubuntu builds to S3, but let the website keep displaying only the `os=linux` build, that'd be ideal. The push to S3 worked as expected, btw, [here is the noble AppImage](http://releases.gitbutler.com/releases/nightly/0.5.1741-2685/linux-noble/x86_64/GitButler_Nightly_0.5.1741_amd64.AppImage.tar.gz), btw.

If that results in anything more than minutes of additional work, I say we just drop the Ubuntu 24.04 build for now and revisit it later.

Byron commented:
As this PR approaches a merge, could we rebuild it into a couple of commits on top of `master`, without the `master` merges? As I keep making test releases from here, I'd have to rebase it on top of `master` before doing so and in the past, I couldn't do that (but would instead have had to merge `master` again).

Regarding the Ubuntu-24 build, I think we can easily remove it for now, wait for the backend fixes, and setup another PR that re-adds them, as only then we have a chance to make them show up on the website anyway.

I'd love to merge this long-standing monster :D.

slarse committed:
Replace Ubuntu 24.04 builds with Ubuntu 22.04

GitButler builds fine on Ubuntu 22.04, and building on this older system
should provide an overall wider range of compatibility with older
distros due to the older GLIBC version (2.35, as opposed to 2.38 on
Ubuntu 24.04).

There are known forward-compatibility issues with the AppImage built on
22.04, just like there are known backward-compatibility issues with the
AppImage built on Ubuntu 24.04. As we're demoting the AppImage to
experimental anyway, this is acceptable.
